python/ray/experimental/dag/dag_node.py

```diff
@@ -84,47 +83,51 @@ def execute(self, *args, **kwargs) -> Union[ray.ObjectRef, ray.actor.ActorHandle
         """Execute this DAG using the Ray default executor."""
         return self.apply_recursive(lambda node: node._execute_impl(*args, **kwargs))
 
-    def _get_toplevel_child_nodes(self) -> Set["DAGNode"]:
-        """Return the set of nodes specified as top-level args.
+    def _get_toplevel_child_nodes(self) -> List["DAGNode"]:
+        """Return the list of nodes specified as top-level args.
 
         For example, in `f.remote(a, [b])`, only `a` is a top-level arg.
 
-        This set of nodes are those that are typically resolved prior to
+        This list of nodes are those that are typically resolved prior to
         task execution in Ray. This does not include nodes nested within args.
         For that, use ``_get_all_child_nodes()``.
         """
 
-        children = set()
+        children = []
         for a in self.get_args():
             if isinstance(a, DAGNode):
-                children.add(a)
```

Comment:
Since python3.6, `set` is ordered. New duplicated objects would not change the order in the set. Why replace it with a list?

Reply:
@suquark for the same objects, yes. But consider the following case:
```
node1 = f.bind()
node2 = f.bind()

s = set()
s.add(node1)
s.add(node2)
```

running above multiple times, you will get s as either (node1, node2) or (node2, node1), as the hash key of the object changed thus the order of the objects are not guaranteed with same code re-run.

Reply:
thanks for the answer! just noticed that this has been answered above. maybe add some comments in the code explaining that?

Reply:
@suquark sure! will do!

Reply:
@suquark from the link i saw a comment "@user474491 Unlike dict, set in Python 3.7+ unfortunately does not preserve order" does it mean CPython 3.6+ supports ordered `dict()` but not for `set()` ?

Reply:
@jiaodong with your comment I just realized that I misunderstood @suquark 's initial post.  And I think @suquark is right if we have **Insertion Ordered** set.  But I think the set implementation is not yet there as you pointed out.

Reply:
The good news here is the function this file touched wasn't on critical path of Ray DAG execution, only if we want to introspect its structure :) But in future development I will keep this topic in mind -- I think have efficient but also ordering guarantee is favorable on the execution path of Ray DAG, it also makes traversal output order deterministic .